Scripts/Runtime/Entities/Lifecycle/JobInteraction/EntitySpawner.cs

```diff
@@ -0,0 +1,342 @@
+using Anvil.Unity.DOTS.Util;
+using System;
+using System.Diagnostics;
+using Unity.Burst;
+using Unity.Collections;
+using Unity.Collections.LowLevel.Unsafe;
+using Unity.Entities;
+
+namespace Anvil.Unity.DOTS.Entities
+{
+    /// <summary>
+    /// Helper struct to allow for spawning <see cref="Entity"/> and making associated structural and initialization
+    /// changes while in parallel bursted jobs.
+    /// </summary>
+    [BurstCompatible]
+    public readonly struct EntitySpawner : IEquatable<EntitySpawner>
+    {
+        public static bool operator ==(EntitySpawner lhs, EntitySpawner rhs)
+        {
+            return lhs.ID == rhs.ID;
+        }
+
+        public static bool operator !=(EntitySpawner lhs, EntitySpawner rhs)
+        {
+            return !(lhs == rhs);
+        }
+        
+        private const int UNSET_THREAD_INDEX = -1;
+        
+        /// <summary>
+        /// The <see cref="SpawnerID"/> to identify this instance globally in the application
+        /// </summary>
+        public readonly SpawnerID ID;
+        
+        /// <summary>
+        /// The Thread Index this instance is operating on.
+        /// </summary>
+        [NativeSetThreadIndex] public readonly int ThreadIndex;
+        
+        
+        [NativeDisableContainerSafetyRestriction] private readonly EntityCommandBuffer m_ECB;
+        [NativeDisableContainerSafetyRestriction] private readonly EntityCommandBuffer.ParallelWriter m_ECBWriter;
+        [ReadOnly] private readonly NativeParallelHashMap<long, EntityArchetype> m_ArchetypeLookup;
+        [ReadOnly] private readonly NativeParallelHashMap<long, Entity> m_PrototypeLookup;
+        
+        
+        internal EntitySpawner(
+            SpawnerID id,
+            EntityCommandBuffer ecb,
+            NativeParallelHashMap<long, EntityArchetype> archetypeLookup,
+            NativeParallelHashMap<long, Entity> prototypeLookup)
+        {
+            ID = id;
+            m_ECB = ecb;
+            m_ECBWriter = m_ECB.AsParallelWriter();
```

Comment:
Is there any overhead associated with always asking for a parallel writer?
In most TaskDriver scenarios the writer won't be written to in parallel.

Reply:
There is probably some small amount of overhead just in terms of making the different lanes that a parallel writer can write to but that's a necessary evil.

With `EntitySpawner`s being included in jobs to directly spawn, we have to allow for parallel jobs to write otherwise we'd be constraining all jobs to execute on one thread only.

> In most TaskDriver scenarios the writer won't be written to in parallel.

I'm not sure that's true? All TaskDriver jobs are designed to go wide with most using `BatchStrategy.MaximizeChunk` so with low volume it MIGHT only execute on one thread, but with more than one chunk worth of data it will definitely go on other threads.

Reply:
> With EntitySpawners being included in jobs to directly spawn, we have to allow for parallel jobs to write otherwise we'd be constraining all jobs to execute on one thread only.

We couldn't have an opt in to a parallel writer? Most jobs that I've written so far the spawn are running on a single thread. They tend to be low volume.

> I'm not sure that's true? All TaskDriver jobs are designed to go wide with most using BatchStrategy.MaximizeChunk so with low volume it MIGHT only execute on one thread, but with more than one chunk worth of data it will definitely go on other threads.

Are they? Almost all the Task Drivers I've written use `Schedule` for the jobs not `ScheduleParallel` to avoid clobbering EPD entries. The BatchStrategy is irrelevant in that situation I thought.